Vendor JSON

denrase opened the pull request:
## :scroll: Description
<!--- Describe your changes in detail -->

This PR will be the baseline where further PRs will be merged into. Therefore it will stay in draft for a longer period of time.

### Vendor

- Add `gson` `JSONReader` and `JSONWriter` with needed files in `vendor/gson` folder
  - Document tag, commit hast and changes of added files
  - Add license in `vendor/gson`
  - Exclude `vendor` files from usual checks so we can keep the source/tests the same.
    - Easier to see updates when importing new files and also no risk of different behaviour.

### Serialization

- Adds `JsonSerializable` that models can implement to write json to a stream.
- Adds `JSONReader` subclass `JsonObjectWriter`
  - Exposes method to call `JsonObjectSerializer`
    - Writes primitives, objects and serializable objects to the json stream.
    - Supports: `null`, `boolean`, `int`, `long`, `double`, `String`, `Map`, `Collection`, `Array`, `JsonSerializable`
    - Unknown objects are serialized with a string placeholder until we support reflection.
- The `JsonSerializable ` interface provides access to a `unknown` map which is serialized using `JsonObjectSerializer `
- Exposes `JsonSerializable` as public API.

### Deserialization

- Adds `JsonDeserializer` that models can implement to read json to a stream.
- Adds `JSONWriter` subclass `JsonObjectReader`.
  - Exposes method to call `JsonObjectDeserializer`.
    - Reads from the json stream and converts it to Java primitives & objects.
    - Supports: `null`, `boolean`, `int`, `long`, `double`, `String`, `Map`, `List`
- The `JsonDeserializer` will take read all unknown fields and set then on the the deserialized object.

### Example

- `UserFeedback` implements serialization/deserialization using the new classes to illustrate the usage.

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Relates to #1537
Resolves #1733
Resolves #1497

- Evaluate approach and plan further steps.

## :green_heart: How did you test it?

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [ ] I updated the docs if needed
- [ ] No breaking changes


## :crystal_ball: Next steps

This PR will stay open until everything is implemented.


- [x] Implement date mapping and fallbacks. [#1634](https://github.com/getsentry/sentry-java/pull/1634)
- [x] Implement serialization/deserialization for all models.
  - [x] App [#1634](https://github.com/getsentry/sentry-java/pull/1634)
  - [x] Browser [#1635](https://github.com/getsentry/sentry-java/pull/1635)
  - [x] Device [#1675](https://github.com/getsentry/sentry-java/pull/1675)
  - [x] OperatingSystem [#1671](https://github.com/getsentry/sentry-java/pull/1671)
  - [x] Runtime [#1672](https://github.com/getsentry/sentry-java/pull/1672)
  - [x] Gpu [#1673](https://github.com/getsentry/sentry-java/pull/1673)
  - [x] SpanContext [#1674](https://github.com/getsentry/sentry-java/pull/1674)
  - [x] SentryBaseEvent [#1712](https://github.com/getsentry/sentry-java/pull/1712)
    - [x] Contexts [#1712](https://github.com/getsentry/sentry-java/pull/1712)
    - [x] SdkVersion [#1688](https://github.com/getsentry/sentry-java/pull/1688)
      - [x] SentryPackage [#1688](https://github.com/getsentry/sentry-java/pull/1688)
    - [x] Request [#1688](https://github.com/getsentry/sentry-java/pull/1688)
    - [x] User [#1688](https://github.com/getsentry/sentry-java/pull/1688)
    - [x] Breadcrumb [#1688](https://github.com/getsentry/sentry-java/pull/1688)
      - [x] SentryLevel [#1688](https://github.com/getsentry/sentry-java/pull/1688)
  - [x] SentryEvent [#1716](https://github.com/getsentry/sentry-java/pull/1716)
    - [x] SentryThread [#1686](https://github.com/getsentry/sentry-java/pull/1686)
      - [x] SentryStackTrace [#1686](https://github.com/getsentry/sentry-java/pull/1686)
        - [x] SentryStackFrame [#1686](https://github.com/getsentry/sentry-java/pull/1686)
    - [x] SentryException [#1687](https://github.com/getsentry/sentry-java/pull/1687)
      - [x] Mechanism [#1687](https://github.com/getsentry/sentry-java/pull/1687)
    - [x] DebugMeta [#1685](https://github.com/getsentry/sentry-java/pull/1685)
      - [x] DebugImage [#1685](https://github.com/getsentry/sentry-java/pull/1685)
      - [x] SdkInfo [#1685](https://github.com/getsentry/sentry-java/pull/1685)
  - [x] SentryTransaction [#1720](https://github.com/getsentry/sentry-java/pull/1720)
    - [x] SentrySpan [#1720](https://github.com/getsentry/sentry-java/pull/1720)
    - [x] MeasurementValue [#1720](https://github.com/getsentry/sentry-java/pull/1720)
  - [x] SentryEnvelope [#1722](https://github.com/getsentry/sentry-java/pull/1722)
    - [x] SentryEnvelopeHeader [#1722](https://github.com/getsentry/sentry-java/pull/1722)
    - [x] SentryEnvelopeItemHeader [#1722](https://github.com/getsentry/sentry-java/pull/1722)
      - [x] SentryItemType [#1722](https://github.com/getsentry/sentry-java/pull/1722)
  - [x] Session [#1724](https://github.com/getsentry/sentry-java/pull/1724)
    - [x] State [#1724](https://github.com/getsentry/sentry-java/pull/1724)
  - [x] TraceState [#1736](https://github.com/getsentry/sentry-java/pull/1736)
    - [x] TraceStateHeader [#1736](https://github.com/getsentry/sentry-java/pull/1736)
- [x] Pull in main and update models
- [x] Use new serialization approach globally [#1744](https://github.com/getsentry/sentry-java/pull/1744)
- [x] Remove previous interfaces and classes
  - [x] Remove IUnknownPropertiesConsumer [#1749](https://github.com/getsentry/sentry-java/pull/1749)
  - [x] Remove `Gson` Dependency [#1744](https://github.com/getsentry/sentry-java/pull/1744)
- [x] Keep track of items for migration docs and changelog

bruno-garcia reviewed (commented):
Would be nice to add a README.md to `vendor/gson` with a link to where we got it (their github I assume? but more importantly, the version (and commit sha), so we can easily diff/check if we need to update things.

Alternatively just add a comment to the top of the class:

```

https://github.com/google/gson/blob/f319c1b8e5ac1135ab253513f91d5ece6719cdf7/gson/src/main/java/com/google/gson/stream/JsonWriter.java
```

bruno-garcia commented on sentry/src/main/java/io/sentry/vendor/gson/stream/JsonScope.java:
```diff
@@ -0,0 +1,79 @@
+ * limitations under the License.
+ */
+
+// Source: https://github.com/google/gson
```
Oh I missed this.

denrase commented:
@bruno-garcia @marandaneto @maciejwalkowiak 
Would love some feedback from you. Basically, I brought in the GSON files we need and added interfaces for serialization/deserialization. Also implemented serialize/deserialize for arbitrary payloads, which is already in use for the `unknown` "bag" so we can preserve everything.

marandaneto commented on sentry/src/main/java/io/sentry/JsonObjectReader.java:
```diff
@@ -0,0 +1,56 @@
+    try {
+      unknown.put(name, nextObjectOrNull());
+    } catch (Exception exception) {
+      String message = "Error deserializing unknown key \"" + name + "\"";
+      logger.log(SentryLevel.ERROR, message, exception);
```
iirc `.log` already has an argument that allows you to use something similar to String.format, eg `"message: %s", message`

marandaneto commented on sentry/src/main/java/io/sentry/JsonObjectWriter.java:
```diff
@@ -0,0 +1,30 @@
+
+  public JsonObjectWriter(Writer out) {
+    super(out);
+    jsonObjectSerializer = new JsonObjectSerializer();
```
could be init. during declaration

marandaneto commented on sentry/src/main/java/io/sentry/UserFeedback.java:
```diff
@@ -115,4 +121,104 @@ public String toString() {
+      reader.endObject();
+
+      if (sentryId == null) {
+        String message = "Missing required field \"" + JsonKeys.EVENT_ID + "\"";
+        Exception exception = new IllegalStateException(message);
+        logger.log(SentryLevel.ERROR, message, exception);
```
similar to https://github.com/getsentry/sentry-java/pull/1554/files#r670436430

denrase replied:
Re-throwing the exception with the message in ctor.

denrase commented:
@marandaneto @maciejwalkowiak Think we are ready for an alpha release, i updated the changelog with the largest changes that this PR introduces.

bruno-garcia commented:
I'll do the honours then :D
